File: kotlin/src/Day02.kt
fun main() {
    fun getSumOfValues(input: List<String>, filter: String): Int {
        return input.filter { x -> x.contains(filter) }.sumOf { x -> x.substring(filter.length + 1).toInt() }
    }

    fun part1(input: List<String>): Int {
        val horizontalPosition = getSumOfValues(input, "forward")
        val down = getSumOfValues(input, "down")
        val up = getSumOfValues(input, "up")
        val depth = down - up
        return horizontalPosition * depth
    }

    fun part2(input: List<String>): Int {
        val horizontalPosition = getSumOfValues(input, "forward")
        var depth = 0;
        var aim = 0;
        input.forEach{
            val parts = it.split(' ')
            when (parts[0]) {
                "forward" -> depth += aim * parts[1].toInt()
                "down" -> aim += parts[1].toInt()
                "up" -> aim -= parts[1].toInt()
            }
        }
        return horizontalPosition * depth
    }

    val testInput = readInput("Day02_test")
    val input = readInput("Day02")

    println(part1(testInput))
    check(part1(testInput) == 150)
    println(part1(input))

    println(part2(testInput))
    check(part2(testInput) == 900)
    println(part2(input))
}

File: typescript/src/day02.ts
import {Utils} from "./utils";

class Operation {
    direction: string;
    amount: number
}

function getSumOfValues(operations: Operation[], direction: string): number {
    return operations.filter(x => x.direction === direction)
        .map(x => x.amount)
        .reduce((a, b) => a + b, 0);
}

function part1(input: string[]): number {
    let operations = input.map(x => {
        let splitString = x.split(' ');
        let direction: string = splitString[0];
        let amount: number = +splitString[1];
        return {direction, amount} as Operation;
    });
    let horizontalPosition = getSumOfValues(operations, "forward");
    let down = getSumOfValues(operations, "down");
    let up = getSumOfValues(operations, "up");
    let depth = down - up;
    return horizontalPosition * depth;
}

function part2(input: string[]): number {
    let operations = input.map(x => {
        let splitString = x.split(' ');
        let direction: string = splitString[0];
        let amount: number = +splitString[1];
        return {direction, amount} as Operation;
    });
    let horizontalPosition = 0
    let depth = 0
    let aim = 0
    for (let operation of operations) {
        switch (operation.direction) {
            case "forward":
                horizontalPosition += operation.amount;
                depth += aim * operation.amount;
                break;
            case "down":
                aim += operation.amount;
                break;
            case "up":
                aim -= operation.amount;
                break;
        }
    }
    return horizontalPosition * depth;
}

let testInput = Utils.readInput("Day02_test");
let input = Utils.readInput("Day02");

console.assert(part1(testInput) === 150);
console.log(part1(input));

console.assert(part2(testInput) === 900);
console.log(part2(input));
